libraries/AP_ExternalAHRS/AP_ExternalAHRS_AdvancedNavigation.cpp

```diff
@@ -0,0 +1,1869 @@
+
+    if (gnss_fix_type > AN_RawGnss::AN_GnssFixType::NONE)
+    {
+        status.flags.horiz_vel = true;
+        status.flags.horiz_pos_rel = true;
+        status.flags.horiz_pos_abs = true;
+        status.flags.pred_horiz_pos_rel = true;
+        status.flags.pred_horiz_pos_abs = true;
+        status.flags.using_gps = true;
+    }
+
+    if (gnss_fix_type > AN_RawGnss::AN_GnssFixType::FIX_2D)
+    {
+        status.flags.gps_quality_good = true;
+    }
+}
+
+void AP_ExternalAHRS_AdvancedNavigation::update()
+{
+    if (!get_packets())
+    {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "ExternalAHRS: Error Receiving Packets");
+    }
+}
+
+// calculate the EKF variances
+bool AP_ExternalAHRS_AdvancedNavigation::get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const
+{
+    const auto *velocity_standard_deviation = reinterpret_cast<const AN_VelocityStandardDeviation *>(&last_velocity_standard_deviation_packet.payload); 
+    const auto *raw_gnss = reinterpret_cast<const AN_RawGnss *>(&last_raw_gnss_packet.payload); 
+    
+    velVar = velocity_standard_deviation->get_horizontal_velocity_accuracy() * vel_gate_scale;
+    posVar = raw_gnss->get_horizontal_position_accuracy() * pos_gate_scale;
+    hgtVar = raw_gnss->get_vertical_position_accuracy() * hgt_gate_scale;
+    tasVar = 0;
+    magVar = Vector3f{0,0,0};
+    
+    return true;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// AP_ExternalAHRS_AdvancedNavigation private methods
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+void AP_ExternalAHRS_AdvancedNavigation::update_thread(void)
+{
+    auto tstart = AP_HAL::millis();
+    unsigned attempts = 0;
+
+    // open the port in the thread
+    _uart->begin(_baudrate);
+
+    while (true)
+    {
+        auto now = AP_HAL::millis();
+        auto init = initialised();
+
+        if (now - tstart >= AN_TIMEOUT)
+        {
+            // If unable to initialise then send errors to GCS periodically
+            if (!init && attempts > AN_MIN_CONNECTION_ATTEMPTS)
+            {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "ExternalAHRS: Advanced Navigation Device Unresponsive");
+            }
+
+            // Request device information periodically
+            if (!configure_device())
+            {
+                GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "ExternalAHRS: Request Data Error");
+            }
+
+            attempts++;
+            tstart = now;
+        }
+
+        hal.scheduler->delay_microseconds(AN_UPDATE_THREAD_PERIOD);
+
+        // Send Aiding data to the device
+        if (init && option_is_set(AP_ExternalAHRS::OPTIONS::ARSP_AID))
+        {
+            if (!send_airspeed_aiding())
+            {
+                GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "ExternalAHRS: Unable to send airspeed aiding.");
```

Comment:
please review all GCS_SEND_TEXT() calls - this and several others could flood the telemetry link. eg. if airspeed went unhealthy then it would flood the link, possibly making control of the vehicle impossible